src/censored.jl

```diff
@@ -0,0 +1,168 @@
+"""
+    censored(d::UnivariateDistribution, l::Real, u::Real)
+
+Censor a univariate distribution `d` to the interval `[l, u]`.
+
+The lower bound `l` can be finite or `-Inf` and the upper bound `u` can be finite or
+`Inf`. The function throws an error if `l > u`.
+
+The function falls back to constructing a [`Censored`](@ref) wrapper.
+
+# Implementation
+
+To implement a specialized censored form for distributions of type `D`, the method
+`censored(d::D, l::T, u::T) where {T <: Real}` should be implemented.
+"""
+function censored(d::UnivariateDistribution, l::Real, u::Real)
+    return censored(d, promote(l, u)...)
+end
+
+function censored(d::UnivariateDistribution, l::T, u::T) where {T <: Real}
+    l <= u || error("the lower bound must be less or equal than the upper bound")
+    Censored(d, promote(l, u)...)
+end
+
+"""
+    Censored
+
+Generic wrapper for a censored distribution.
+"""
+struct Censored{D<:UnivariateDistribution, S<:ValueSupport, T <: Real} <: UnivariateDistribution{S}
+    uncensored::D      # the original distribution (uncensored)
+    lower::T      # lower bound
+    upper::T      # upper bound
+    function Censored(d::UnivariateDistribution, l::T, u::T) where {T <: Real}
+        new{typeof(d), value_support(typeof(d)), T}(d, l, u)
+    end
+end
+
+params(d::Censored) = tuple(params(d.uncensored)..., d.lower, d.upper)
+partype(d::Censored) = partype(d.uncensored)
+Base.eltype(::Type{Censored{D, S, T} } ) where {D, S, T} = T
+
+### range and support
+
+islowerbounded(d::Censored) = islowerbounded(d.uncensored) || isfinite(d.lower)
+isupperbounded(d::Censored) = isupperbounded(d.uncensored) || isfinite(d.upper)
+
+minimum(d::Censored) = max(minimum(d.uncensored), d.lower)
+maximum(d::Censored) = min(maximum(d.uncensored), d.upper)
+
+function insupport(d::Censored{D,<:Union{Discrete,Continuous}}, x::Real) where {D<:UnivariateDistribution}
+    return d.lower <= x <= d.upper && insupport(d.uncensored, x)
+end
+
+### evaluation
+
+quantile(d::Censored, p::Real) = clamp(quantile(d.uncensored, p), d.lower, d.upper)
+
+function pdf(d::Censored, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    return if x == lower
+        cdf(d0, x)
+    elseif x == upper
+        uccdf = ccdf(d0, x)
+        if value_support(typeof(d)) == Discrete
+            uccdf - pdf(d0, x)
+        else
+            uccdf
+        end
+    else
+        result = pdf(d0, x)
+        ifelse(lower < x < upper, result, zero(result))
```

Comment:
Is there any specific reason for `ifelse`?
```suggestion
        lower < x < upper ? result : zero(result)
```